# Predictive Maintenance Intelligence Platform

## 01 - Data Understanding

**Objective:** Understand the dataset structure, data quality, and feature information before preprocessing and model building.

In [2]:
#Import required libraries
import pandas as pd
import numpy as np 

In [3]:
#Load the dataset
df = pd.read_csv("../data/raw/AI4I-PMDI.csv")

In [4]:
#Display the first five rows of the dataset
df.head()

,UDI,Date,System,Control,Product ID,Type,Air temperature (K),Process temperature (K),Rotational speed (rpm),Torque (Nm),Tool wear (min),Diagnostic
0,1,15/04/2014 11:56,0,C,M14860,M,NaN,NaN,NaN,42.8,0.0,No failure
1,2,12/04/2014 16:09,0,A,L47181,L,298.2,308.7,1408.0,NaN,NaN,No failure
2,3,13/04/2014 01:13,0,A,L47182,L,298.1,308.5,1498.0,NaN,NaN,No failure
3,4,24/07/2014 20:35,0,B,L47183,L,NaN,NaN,1433.0,39.5,NaN,No failure
4,5,22/07/2014 01:31,0,C,L47184,L,NaN,NaN,NaN,40.0,9.0,No failure


### Observations

- The dataset loaded successfully.
- The dataset contains both numerical and categorical features.
- Missing values are already visible in multiple sensor columns.
- The target variable is `Diagnostic`.
- Time-related (`Date`) and identifier (`UDI`, `Product ID`) columns are present.

In [6]:
#Check the dimensions of the dataset
df.shape

(10000, 12)

### Observation

- The dataset contains **10,000 records** and **12 columns**.
- The dataset size is suitable for exploratory data analysis and machine learning.

In [8]:
#Display all column names
df.columns 

Index(['UDI', 'Date', 'System', 'Control', 'Product ID', 'Type',
       'Air temperature (K)', 'Process temperature (K)',
       'Rotational speed (rpm)', 'Torque (Nm)', 'Tool wear (min)',
       'Diagnostic'],
      dtype='object')

In [9]:
#Display dataset information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   UDI                      10000 non-null  int64  
 1   Date                     10000 non-null  object 
 2   System                   10000 non-null  int64  
 3   Control                  10000 non-null  object 
 4   Product ID               10000 non-null  object 
 5   Type                     10000 non-null  object 
 6   Air temperature (K)      3437 non-null   float64
 7   Process temperature (K)  3437 non-null   float64
 8   Rotational speed (rpm)   6679 non-null   float64
 9   Torque (Nm)              6563 non-null   float64
 10  Tool wear (min)          3321 non-null   float64
 11  Diagnostic               10000 non-null  object 
dtypes: float64(5), int64(2), object(5)
memory usage: 937.6+ KB


In [10]:
#Display the statistical summary of numerical columns 
df.describe()

,UDI,System,Air temperature (K),Process temperature (K),Rotational speed (rpm),Torque (Nm),Tool wear (min)
count,10000.00000,10000.000000,3437.000000,3437.000000,6679.000000,6563.000000,3321.000000
mean,5000.50000,58.840800,300.064358,310.033081,1541.242402,39.965641,110.524240
std,2886.89568,34.206948,2.016596,1.477490,186.509386,10.164474,64.563871
min,1.00000,0.000000,295.400000,305.800000,1168.000000,3.800000,0.000000
25%,2500.75000,29.000000,298.400000,308.900000,1423.000000,33.100000,55.000000
50%,5000.50000,59.000000,300.100000,310.100000,1502.000000,39.900000,110.000000
75%,7500.25000,88.000000,301.600000,311.100000,1613.500000,46.700000,167.000000
max,10000.00000,119.000000,304.300000,313.800000,2886.000000,76.600000,253.000000


In [11]:
#Display the statistical summary of categorical columns 
df.describe(include='object')

,Date,Control,Product ID,Type,Diagnostic
count,10000,10000,10000,10000,10000
unique,9986,3,10000,3,6
top,01/07/2015 20:20,A,M14860,L,No failure
freq,2,3437,1,6000,9652


In [21]:
#Check missing values in each column
df.isnull().sum()

UDI                           0
Date                          0
System                        0
Control                       0
Product ID                    0
Type                          0
Air temperature (K)        6563
Process temperature (K)    6563
Rotational speed (rpm)     3321
Torque (Nm)                3437
Tool wear (min)            6679
Diagnostic                    0
dtype: int64

In [25]:
#Calculate the percentage of missing values 
missing_percentage=(df.isnull().sum()/len(df))*100
missing_percentage.sort_values(ascending=False)

Tool wear (min)            66.79
Air temperature (K)        65.63
Process temperature (K)    65.63
Torque (Nm)                34.37
Rotational speed (rpm)     33.21
UDI                         0.00
Date                        0.00
System                      0.00
Control                     0.00
Product ID                  0.00
Type                        0.00
Diagnostic                  0.00
dtype: float64

In [27]:
#check for duplicated rows
df.duplicated().sum()

0

# Data Understanding Summary

## Key Findings

- The dataset contains **10,000 records** and **12 columns**.
- It consists of numerical, categorical, identifier, and target features.
- Five sensor-related columns contain missing values, with some exceeding **65% missing data**.
- No duplicate records were found.
- The `Date` column is currently stored as an object and may require conversion to datetime during preprocessing.
- The `Product ID` column appears to be a unique identifier and will be evaluated during feature selection.
- The target variable (`Diagnostic`) contains multiple categories, with **"No failure"** being the most frequent.